vm/src/builtins.rs

```diff
@@ -13,6 +13,78 @@ use super::pyobject::{
 };
 use super::vm::VirtualMachine;
 
+macro_rules! arg_check {
+    ( $vm: ident, $args:ident ) => {
+        // Zero-arg case
+        if $args.args.len() != 0 {
+            return Err($vm.new_type_error(format!(
+                "Expected no arguments (got: {})", $args.args.len())));
+        }
+    };
+    ( $vm: ident, $args:ident, $( ($arg_name:ident, $arg_type:expr) ),* ) => {
+        let mut expected_args = vec![];
+        let mut arg_count = 0;
+
+        $(
+            if arg_count >= $args.args.len() {
+                // TODO: Report the number of expected arguments
+                return Err($vm.new_type_error(format!(
+                    "Expected more arguments (got: {})", $args.args.len())));
+            }
+            expected_args.push((arg_count, stringify!($arg_name), $arg_type));
+            let $arg_name = &$args.args[arg_count];
+            #[allow(unused_assignments)]
+            {
+                arg_count += 1;
+            }
+        )*
+
+        if $args.args.len() != expected_args.len() {
+            return Err($vm.new_type_error(format!(
+                "Expected {} arguments (got: {})",
+                expected_args.len(), $args.args.len())));
+        };
+
+        for (arg, (arg_position, arg_name, expected_type)) in $args.args.iter().zip(expected_args.iter()) {
+            let expected_type = match expected_type {
+                'c' => "code object",
```

Comment:
Are the one letter abbreviations the best thing here? Would full type names be clearer?

Reply:
We could use python type names, such as `str`, `int`, `code`, etc..